contracts/multisig-improved/src/action_types/external_module.rs

```diff
@@ -0,0 +1,155 @@
+            esdt_payments: PaymentsVec<Self::Api>,
+        ) -> bool;
+    }
+}
+
+#[multiversx_sc::module]
+pub trait ExternalModuleModule:
+    crate::common_functions::CommonFunctionsModule + crate::state::StateModule
+{
+    #[endpoint(enableModule)]
+    fn enable_module(&self, module: ManagedAddress) {
+        self.set_module_status_common(module, ENABLED);
+    }
+
+    #[endpoint(disableModule)]
+    fn disable_module(&self, module: ManagedAddress) {
+        self.set_module_status_common(module, DISABLED);
+    }
+
+    #[endpoint(addAdditionalAllowedAddresses)]
+    fn add_additional_allowed_addresses(
+        &self,
+        module: ManagedAddress,
+        addresses: MultiValueEncoded<ManagedAddress>,
+    ) {
+        let module_id = self.module_id().get_id_non_zero(&module);
+        self.require_module_owner_caller(module_id);
+
+        let mut mapper = self.additional_allowed_addresses(module_id);
+        for addr in addresses {
+            let (user_id, user_role) = self.get_id_and_role(&addr);
+            user_role.require_can_propose::<Self::Api>();
+
+            let _ = mapper.insert(user_id);
+        }
+    }
+
+    #[endpoint(removeAdditionalAllowedAddresses)]
+    fn remove_additional_allowed_addresses(
+        &self,
+        module: ManagedAddress,
+        addresses: MultiValueEncoded<ManagedAddress>,
+    ) {
+        let module_id = self.module_id().get_id_non_zero(&module);
+        self.require_module_owner_caller(module_id);
+
+        let mut mapper = self.additional_allowed_addresses(module_id);
+        for addr in addresses {
+            let user_id = self.user_ids().get_id_non_zero(&addr);
+            let removed = mapper.swap_remove(&user_id);
+            require!(removed, "Address not in list");
+        }
+    }
+
+    #[view(getModuleStatus)]
+    fn get_module_status(&self, module: ManagedAddress) -> ModuleStatus {
+        let module_id = self.module_id().get_id_non_zero(&module);
+
+        self.module_status(module_id).get()
+    }
+
+    fn set_module_status_common(&self, module: ManagedAddress, status: ModuleStatus) {
+        let module_id = self.module_id().get_id_non_zero(&module);
+        self.require_module_owner_caller(module_id);
+
+        self.module_status(module_id).set(status);
+    }
+
+    fn require_module_owner_caller(&self, module_id: ModuleId) {
+        let caller = self.blockchain().get_caller();
+        let module_owner = self.module_owner(module_id).get();
+        let (caller_id, caller_role) = self.get_id_and_role(&caller);
+        caller_role.require_can_propose::<Self::Api>();
+
+        require!(
+            caller_id == module_owner,
+            "Only module owner can call this function"
+        );
+    }
+
+    fn can_execute_action(
+        &self,
+        sc_address: ManagedAddress,
+        egld_value: BigUint,
+        esdt_payments: PaymentsVec<Self::Api>,
+    ) -> bool {
+        let module_id = self.module_id().get_id(&sc_address);
```

Comment:
There can be more than one modules, so we should have a list and iterate over can_execute